lib/chef/resource/windows_feature_powershell.rb

```diff
@@ -57,77 +58,170 @@ class WindowsFeaturePowershell < Chef::Resource
+        end
+
+        # @return [Array] features the user has requested to remove which need removing
+        def features_to_remove
+          # the intersection of the features to remove & enabled features are what needs removing
+          @remove ||= new_resource.feature_name & node["powershell_features_cache"]["enabled"]
+        end
+
+        # @return [Array] features the user has requested to delete which need deleting
+        def features_to_delete
+          # the intersection of the features to remove & enabled/disabled features are what needs removing
+          @remove ||= begin
+            all_available = node["powershell_features_cache"]["enabled"] +
+              node["powershell_features_cache"]["disabled"]
+            new_resource.feature_name & all_available
           end
         end
 
-        def available?
-          @available ||= begin
-            # @todo remove this when we're ready to drop windows 8/2012
-            cmd = if node["os_version"].to_f < 6.2
-                    powershell_out("Import-Module ServerManager; @(Get-WindowsFeature #{new_resource.feature_name.join(',')}).count", timeout: new_resource.timeout)
-                  else
-                    powershell_out("@(Get-WindowsFeature #{new_resource.feature_name.join(',')} | ?{$_.InstallState -ne \'Removed\'}).count", timeout: new_resource.timeout)
-                  end
-            cmd.stderr.empty? && cmd.stdout.chomp.to_i > 0
+        # if any features are not supported on this release of Windows or
+        # have been deleted raise with a friendly message. At one point in time
+        # we just warned, but this goes against the behavior of ever other package
+        # provider in Chef and it isn't clear what you'd want if you passed an array
+        # and some features were available and others were not.
+        # @return [void]
+        def fail_if_unavailable
+          all_available = node["powershell_features_cache"]["enabled"] +
+            node["powershell_features_cache"]["disabled"] +
+            node["powershell_features_cache"]["removed"]
+
+          # the difference of desired features to install to all features is what's not available
+          unavailable = (new_resource.feature_name - all_available)
+          raise "The Windows feature#{'s' if unavailable.count > 1} #{unavailable.join(',')} #{unavailable.count > 1 ? 'are' : 'is'} not available on this version of Windows. Run 'Get-WindowsFeature' to see the list of available feature names." unless unavailable.empty?
+        end
+
+        # run Get-WindowsFeature to get a list of all available features and their state
+        # and save that to the node at node.override level.
+        # @return [void]
+        def reload_cached_powershell_data
+          Chef::Log.debug("Caching Windows features available via Get-WindowsFeature.")
+          node.override["powershell_features_cache"] = Mash.new
```

Comment:
This shouldn't go in the node, it should really go into a singleton object, similar in principle to the old YumCache